### PR TITLE
IDEX-4222: Change model of machine & workspace

### PR DESCRIPTION
**Pull request by garagatyi**
Machine model changes:
- MachineState was removed, Machine should be used instead
- Machine states API was removed
- Machine includes MachineConfig as field instead of extending it
- All runtime information of machine is in runtime field of Machine.
- Now API call GET _/api/machine?workspace=ws-id123456_ returns all existing machines of workspace,
  including machines with status CREATING, DESTROYING
- Machine channels was removed from. To get channels clients should check links of WorkspaceConfigDto.

Workspace model changes:
- EnvironmentState was removed
- UsersWorkspace includes WorkspaceConfig as field instead of extending it
- Start tmp workspace API returns RuntimeWorkspace instead of UsersWorkspace
- Machine channels are included in links of MachineConfigDto of WorkspaceConfig. Don't try to find them
  in runtime machines of RuntimeWorkspaceDto.
- Field activeEnvName in RuntimeWorkspace was renamed to envName in accordance with defaultEnv field
  of WorkspaceConfig.

Also:
- add static method cloneDto to DtoFactory
- add getParameter(paramName) method to DTO Link to simplify parameter retrieval

Signed-off-by: Alexander Garagatyi agaragatyi@codenvy.com

@skabashnyuk @vparfonov @evoevodin @dimasnurenko Please, review.

### CODE REVIEW

**Comment by benoitf** [comment]
From what I see now instead of machine.getLimits().getRam() we've to do machine.getConfig().getLimits().getRam()

but for example I don't see any change on https://github.com/eclipse/che/blob/master/dashboard/src/components/api/che-workspace.factory.js#L183 or https://github.com/eclipse/che/blob/master/dashboard/src/app/workspaces/list-workspaces/workspace-item/workspace-item.html#L24

**Comment by garagatyi** [comment]
Yes, I didn't change dashboard.

**Comment by skabashnyuk** [comment]
@gazarenkov ?

File: core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Machine.java [hunk 1]
```diff
@@ -14,10 +14,43 @@
...
+     * Runtime status of the machine
+     */
+    MachineStatus getStatus();
+
+    /**
+     * Runtime information about machine.
+     * <p>
+     * Is available only when {@link #getStatus()} returns {@link MachineStatus#RUNNING}
+     */
+    MachineRuntimeInfo getRuntime();
```
Comment (voievodin) [inline core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Machine.java:58]: Add `@Nullable`

Comment (garagatyi) [inline core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Machine.java:58]: ok

Comment (gazarenkov) [inline core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Machine.java:58]: Move docker properties (which currently are in MachineMetadata) directly to MachineRuntimeInfo
So remove MachineMetadata

**Comment by gazarenkov** [comment]
No more comments except moving docker properties (below).
OK for me.

**Comment by skabashnyuk** [comment]
ok

**Pull request merged**